test/Test/Pos/Util/ModifierSpec.hs

```diff
@@ -0,0 +1,172 @@
+{-# LANGUAGE GADTs #-}
+
+-- | Specification for Pos.Util.Modifier (from 'core')
+
+module Test.Pos.Util.ModifierSpec
+       ( spec
+       ) where
+
+import qualified Data.HashMap.Strict   as HM
+
+import qualified Pos.Util.Modifier     as Core
+import           Universum
+
+import           Test.Hspec            (Spec, describe)
+import           Test.Hspec.QuickCheck (prop)
+import           Test.Pos.Util         (formsMonoid)
+import           Test.QuickCheck       (Property)
+
+spec :: Spec
+spec = describe "MapModifier" $ do
+    describe "Monoid instance" $ do
+        prop
+            "Map modifiers form a monoid instance with the custom binary operation"
+            (mapModifierFormsMonoid @Text @Text)
+    describe "Operations on map" $ do
+        describe "insert" $ do
+            prop
+                "inserting a value and then looking it up returns the inserted value"
+                (insertThenLookup @Text @Text)
+        describe "delete" $ do
+            prop
+                "inserting a value and then deleting it removes it from the map"
+                (insertDeleteThenLookup @Text @Text)
+        describe "keys" $ do
+            prop
+                "getting the keys from a map modifier does not eliminate any from the\
+                \ produced list, unless their value was 'Nothing' or they are in the\
+                \ argument list and are a key in the map modifier simultanously"
+                (allKeysAreInMap @Text @Text)
+        describe "values" $ do
+            prop
+                "getting the values from a map modifier does not eliminate any from the\
+                \ produced list, unless their value is 'Nothing' or if it is the value\
+                \ of a key in the input list that does not belong to the the map modifier"
+                (allValsAreInMap @Text @Text)
```

Comment:
This description is correct, but:
1. Again, I don't like that it mentions that value is `Nothing` for the reasons described above.
2. It doesn't fully describe semantics of `values`. This function returns all values from underlying storage except deleted values, it also overrides values according to insertions into `MapModifier` and it also returns new values inserted into `MapModifier`.